File: 2096-step-by-step-directions-from-a-binary-tree-node-to-another/2096-step-by-step-directions-from-a-binary-tree-node-to-another.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getDirections(self, root: Optional[TreeNode], startValue: int, destValue: int) -> str:
        def findPath(node: TreeNode, value: int, path: list) -> bool:
            if node.val == value:
                return True
            if node.left and findPath(node.left, value, path):
                path.append('L')
            elif node.right and findPath(node.right, value, path):
                path.append('R')
            return path
        startPath = []
        destPath = []
        findPath(root, startValue, startPath)
        findPath(root, destValue, destPath)
        while startPath and destPath and startPath[-1] == destPath[-1]:
            startPath.pop()
            destPath.pop()
        return 'U'*len(startPath) + ''.join(reversed(destPath))

File: 2096-step-by-step-directions-from-a-binary-tree-node-to-another/2096-step-by-step-directions-from-a-binary-tree-node-to-another.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public String getDirections(TreeNode root, int startValue, int destValue) {
        StringBuilder startString = new StringBuilder();
        findPath(root, startValue, startString);
        StringBuilder destString = new StringBuilder();
        findPath(root, destValue, destString);
        startString.reverse();
        destString.reverse();
        int i = 0;
        int maxLen = Math.min(startString.length(), destString.length());
        while(i<maxLen && startString.charAt(i)==destString.charAt(i))i++;
        return "U".repeat(startString.length()-i) + destString.substring(i);
    }
    private boolean findPath(TreeNode node, int val, StringBuilder path){
        if(val == node.val)return true;
        if(node.left!=null && findPath(node.left, val, path)){
            path.append('L');
        }
        else if(node.right!=null && findPath(node.right, val, path)){
            path.append('R');
        }
        return path.length()>0;
    }
}